src/sentry/workflow_engine/models/detector.py

```diff
@@ -111,3 +126,80 @@ def __init__(self, detector: Detector):
+
+    def bulk_get_detector_state(
+        self, state_updates: list[DetectorStateData]
+    ) -> dict[str | None, DetectorState]:
+        group_keys = {update.group_key for update in state_updates}
+        query_filter = Q(
+            detector_group_key__in=[group_key for group_key in group_keys if group_key is not None]
+        )
+        if None in group_keys:
+            query_filter |= Q(detector_group_key__isnull=True)
+
+        return {
+            detector_state.detector_group_key: detector_state
+            for detector_state in self.detector.detectorstate_set.filter(query_filter)
+        }
+
+    def commit_state_update_data(self, state_updates: list[DetectorStateData]):
+        detector_state_lookup = self.bulk_get_detector_state(state_updates)
+        created_detector_states = []
+        updated_detector_states = []
+        dedupe_values = []
+        group_counter_updates = {}
+        for state_update in state_updates:
+            detector_state = detector_state_lookup.get(state_update.group_key)
+            if not detector_state:
+                created_detector_states.append(
+                    DetectorState(
+                        detector_group_key=state_update.group_key,
+                        detector=self.detector,
+                        active=state_update.active,
+                        state=state_update.status,
+                    )
+                )
+            elif (
+                state_update.active != detector_state.active
+                or state_update.status != detector_state.state
+            ):
+                detector_state.active = state_update.active
+                detector_state.state = state_update.status
+                updated_detector_states.append(detector_state)
+
+            dedupe_values.append((state_update.group_key, state_update.dedupe_value))
+            group_counter_updates[state_update.group_key] = state_update.counter_updates
+
+        if created_detector_states:
+            DetectorState.objects.bulk_create(created_detector_states)
+
+        if updated_detector_states:
+            DetectorState.objects.bulk_update(updated_detector_states, ["active", "state"])
+
+        pipeline = get_redis_client().pipeline()
+        if dedupe_values:
+            for group_key, dedupe_value in dedupe_values:
+                pipeline.set(self.build_dedupe_value_key(group_key), dedupe_value, ex=REDIS_TTL)
+
+        if group_counter_updates:
+            for group_key, counter_updates in group_counter_updates.items():
+                for counter_name, counter_value in counter_updates.items():
+                    key_name = self.build_counter_value_key(group_key, counter_name)
+                    if counter_value is None:
+                        pipeline.delete(key_name)
+                    else:
+                        pipeline.set(key_name, counter_value, ex=REDIS_TTL)
+
+        pipeline.execute()
```

Comment:
nit; might be nice to decompose this into an `update_redis` method.

Reply:
Yeah, I think breaking this up more could be nicer